File: app/src/main/java/timothypaetz/com/recyclersectionheader/MainActivity.java
package timothypaetz.com.recyclersectionheader;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import java.util.Collections;
import java.util.List;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(this,
                                                              LinearLayoutManager.VERTICAL,
                                                              false));

        final List<Person> people = getPeople();

        RecyclerSectionItemDecoration sectionItemDecoration =
            new RecyclerSectionItemDecoration(getResources().getDimensionPixelSize(R.dimen.recycler_section_header_height),
                                              false);

        sectionItemDecoration.setCallback(new RecyclerSectionItemDecoration.SectionCallback() {
            @Override
            public boolean isSection(int position) {
                return position == 0
                    || people.get(position)
                             .getLastName()
                             .charAt(0) != people.get(position - 1)
                                                 .getLastName()
                                                 .charAt(0);
            }

            @Override
            public CharSequence getSectionHeader(int position) {
                return people.get(position)
                             .getLastName()
                             .subSequence(0,
                                          1);
            }
        });
        recyclerView.addItemDecoration(sectionItemDecoration);

        recyclerView.setAdapter(new PersonAdapter(getLayoutInflater(),
                                                  people,
                                                  R.layout.recycler_row));
    }

    private List<Person> getPeople() {
        PersonRepo personRepo = new PersonRepo();
        List<Person> people = personRepo.getPeople();
        Collections.sort(people);
        return people;
    }
}

File: app/src/main/java/timothypaetz/com/recyclersectionheader/MainActivity.kt
package timothypaetz.com.recyclersectionheader

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import timothypaetz.com.recyclersectionheader.RecyclerSectionItemDecoration.SectionCallback
import java.util.*

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val recyclerView: RecyclerView = findViewById(R.id.recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(
            this,
            LinearLayoutManager.VERTICAL,
            false
        )
        val presidents = PersonRepo().people.sorted()
        val sectionItemDecoration = RecyclerSectionItemDecoration(
            resources.getDimensionPixelSize(R.dimen.recycler_section_header_height),
            true,
            getSectionCallback(presidents)
        )
        recyclerView.addItemDecoration(sectionItemDecoration)
        recyclerView.adapter = PersonAdapter(
            layoutInflater,
            presidents,
            R.layout.recycler_row
        )
    }

    private fun getSectionCallback(people: List<Person>): SectionCallback {
        return object : SectionCallback {
            override fun isSection(position: Int): Boolean {
                return (position == 0 || people[position].lastName[0] != people[position - 1].lastName[0])
            }

            override fun getSectionHeader(position: Int): CharSequence {
                return people[position].lastName.subSequence(0, 1)
            }
        }
    }
}